### PR TITLE
HSTU: add attention backend dispatcher (auto/cpp) with safe fallback; opt-in gin

### LINKED ISSUE #308: Efficiency on H100 GPU
Hi, thanks for sharing this great work.

I tested this code on ml-1m dataset on a single Nvidia H100 GPU card and it took about 90 minutes to complete 100 epochs. I wonder if this is the expected speed of the code? Thanks.

### PR DESCRIPTION
**Pull request by TeleViaBox**
### Summary
- Add a minimal attention-backend dispatcher for **HSTU** controlled by a new gin flag `HSTU.attn_impl`.
- Supported values today: `"auto"` (default-fallback) and `"cpp"`.  
  Default configs remain unchanged; only users who opt in will switch backend.
- On Hopper + compiled C++ extension we select **cpp**; otherwise we safely fall back to the existing path.
- Fixes #308.

### Changed files
- generative_recommenders/research/modeling/sequential/hstu.py
- generative_recommenders/ops/hstu_attention_dispatch.py
- configs/ml-1m/hstu-sampled-softmax-n128-large-h100.gin

### Reproduction environment (excerpt)
- OS: Ubuntu 22.04
- kernel: 5.15.0-116-generic
- GPU: NVIDIA RTX A6000 (48 GB, SM 8.6)
- torch: 2.5.1+cu124
- CUDA: 12.4
- fbgemm-gpu: 1.0.0+cu124
- commit: ea466f3

### Baseline run (unchanged pipeline)
Command:
```

/usr/bin/time -v bash -lc&#x20;
'CUDA\_VISIBLE\_DEVICES=0 python main.py&#x20;
\--gin\_config\_file=configs/ml-1m/hstu-sampled-softmax-n128-large-final.gin&#x20;
\--master\_port=12345' 2>&1 | tee baseline\_ml1m.log

```

Representative log lines:
```

INFO\:root: rank: 0, batch-stat (train): step 0 (epoch 0 in 33.27s): 8.307612
INFO\:root\:rank 0: eval @ epoch 0 in 32.08s: NDCG\@10 0.0155, NDCG\@50 0.0349, HR\@10 0.0321, HR\@50 0.1225, MRR 0.0175
INFO\:root\:rank 0: eval @ epoch 1 in 32.29s: NDCG\@10 0.0482, NDCG\@50 0.0802, HR\@10 0.0906, HR\@50 0.2394, MRR 0.0453
Elapsed (wall clock) time (h\:mm\:ss or m\:ss): 2:06:14

```

### Patched run (opt-in dispatch)
Command:
```

/usr/bin/time -v bash -lc&#x20;
'CUDA\_VISIBLE\_DEVICES=0 python main.py&#x20;
\--gin\_config\_file=configs/ml-1m/hstu-sampled-softmax-n128-large-h100.gin&#x20;
\--master\_port=12345' 2>&1 | tee patched\_ml1m.log

```

Representative log lines:
```

INFO\:root: rank: 0, batch-stat (train): step 0 (epoch 0 in 30.95s): 8.180079
INFO\:root\:rank 0: eval @ epoch 0 in 38.44s: NDCG\@10 0.0155, NDCG\@50 0.0363, HR\@10 0.0301, HR\@50 0.1290, MRR 0.0180
INFO\:root\:rank 0: eval @ epoch 1 in 28.86s: NDCG\@10 0.0454, NDCG\@50 0.0785, HR\@10 0.0889, HR\@50 0.2434, MRR 0.0423
Elapsed (wall clock) time (h\:mm\:ss or m\:ss): 2:06:49
\[HSTU] Using attention backend: triton/ref (fallback)

```

Notes:
- On this A6000 (sm_86) the dispatcher reports `fallback` as expected; training proceeds unchanged.
- Default behavior for all existing configs remains unchanged.
- Hopper users can enable the C++ backend by installing the extension and setting `HSTU.attn_impl="auto"` (or using the new gin).

### CODE REVIEW

**Comment by TeleViaBox** [comment]
CLA signed, all checks green.
- This PR only adds optional backend dispatch for HSTU (default behavior unchanged).
- On non-Hopper GPUs it safely falls back; baseline/opt-in runs have been validated on A6000 with logs attached.

Kindly review, thanks! ✅
